File: src/drivers/MaestroDriver.ts
import { spawn } from 'child_process';
import * as path from 'path';
import type { EnvironmentConfig } from '../../config/types';

export interface MaestroRunOptions {
  env?: Record<string, string>;
  timeoutMs?: number;
  reinstallDriver?: boolean;
}

export interface MaestroResult {
  success: boolean;
  output: string;
  exitCode: number;
}

export class MaestroDriver {
  constructor(private config: EnvironmentConfig) {}

  async run(flowName: string, opts?: MaestroRunOptions): Promise<MaestroResult> {
    const flowPath = path.resolve(this.config.mobile.flowsDir, `${flowName}.yaml`);
    const timeoutMs = opts?.timeoutMs ?? this.config.mobile.timeoutMs;
    const maestroBin = this.config.mobile.maestroBin ?? 'maestro';
    const stream = process.env.DEBUG_MAESTRO !== '0' && process.env.DEBUG_MAESTRO !== 'false';

    const args = ['test', flowPath];
    if (opts?.reinstallDriver) args.push('--reinstall-driver');
    if (this.config.mobile.deviceId) args.unshift('--device', this.config.mobile.deviceId);
    if (opts?.env) {
      for (const [k, v] of Object.entries(opts.env)) args.push('--env', `${k}=${v}`);
    }

    if (stream) {
      const envSummary = opts?.env ? ' ' + Object.entries(opts.env).map(([k, v]) => `${k}=${v}`).join(' ') : '';
      process.stderr.write(`\n[maestro] ▶ ${flowName}${envSummary}\n`);
    }

    return new Promise((resolve) => {
      let output = '';
      let buffer = '';
      const proc = spawn(maestroBin, args, { env: process.env });

      const onData = (chunk: Buffer) => {
        const text = chunk.toString();
        output += text;
        if (!stream) return;
        buffer += text;
        const lines = buffer.split('\n');
        buffer = lines.pop() ?? '';
        for (const line of lines) {
          if (line.trim()) process.stderr.write(`[maestro]   ${line}\n`);
        }
      };
      proc.stdout.on('data', onData);
      proc.stderr.on('data', onData);

      const timer = setTimeout(() => {
        proc.kill('SIGTERM');
        setTimeout(() => { try { proc.kill('SIGKILL'); } catch {} }, 2000);
        if (stream) process.stderr.write(`[maestro] ✖ ${flowName} — timeout after ${timeoutMs}ms\n`);
        resolve({ success: false, output: output + '\nTimeout exceeded', exitCode: -1 });
      }, timeoutMs);

      proc.on('close', (code) => {
        clearTimeout(timer);
        if (stream && buffer.trim()) process.stderr.write(`[maestro]   ${buffer}\n`);
        const exitCode = code ?? -1;
        const failed = exitCode !== 0 || output.includes('Flow Failed');
        if (stream) process.stderr.write(`[maestro] ${failed ? '✖' : '✔'} ${flowName} (exit=${exitCode})\n`);
        resolve({ success: !failed, output, exitCode });
      });
    });
  }

  async runOrThrow(flowName: string, opts?: MaestroRunOptions): Promise<MaestroResult> {
    const result = await this.run(flowName, opts);
    if (!result.success) {
      // Only retry with --reinstall-driver on real on-device gRPC crashes, NOT on our own timeout
      const driverCrashed = /io\.grpc\.StatusRuntimeException: UNAVAILABLE|Command failed \(tcp:\d+\): closed/.test(result.output);
      if (driverCrashed && !opts?.reinstallDriver) {
        process.stderr.write(`[maestro] ⚠ device driver crashed — retrying with --reinstall-driver\n`);
        const retry = await this.run(flowName, { ...opts, reinstallDriver: true });
        if (retry.success) return retry;
      }
      throw new Error(`Maestro flow "${flowName}" failed:\n${result.output}`);
    }
    return result;
  }

  async takeScreenshot(name: string): Promise<string> {
    const outputDir = 'reports/screenshots';
    const screenshotPath = path.resolve(outputDir, `${name}.png`);
    const deviceArgs = this.config.mobile.deviceId ? ['-s', this.config.mobile.deviceId] : [];
    const adb = process.env.ANDROID_HOME
      ? path.join(process.env.ANDROID_HOME, 'platform-tools', 'adb')
      : 'adb';

    return new Promise((resolve, reject) => {
      const proc = spawn(adb, [...deviceArgs, 'exec-out', 'screencap', '-p'], { env: process.env });
      const chunks: Buffer[] = [];
      proc.stdout.on('data', (chunk: Buffer) => chunks.push(chunk));
      proc.on('close', (code) => {
        if (code === 0 && chunks.length > 0) {
          const fs = require('fs');
          fs.mkdirSync(outputDir, { recursive: true });
          fs.writeFileSync(screenshotPath, Buffer.concat(chunks));
          resolve(screenshotPath);
        } else {
          reject(new Error(`adb screencap failed with code ${code}`));
        }
      });
    });
  }
}

File: java/conductor-core/src/main/java/com/nouhouari/conductor/drivers/MaestroDriver.java
package com.nouhouari.conductor.drivers;

import com.nouhouari.conductor.config.EnvironmentConfig;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

/** Java port of src/drivers/MaestroDriver.ts — shells out to the Maestro CLI. */
public class MaestroDriver {

    public record Result(boolean success, String output, int exitCode) {
    }

    private static final Pattern DRIVER_CRASH =
            Pattern.compile("io\\.grpc\\.StatusRuntimeException: UNAVAILABLE|Command failed \\(tcp:\\d+\\): closed");

    private final EnvironmentConfig config;

    public MaestroDriver(EnvironmentConfig config) {
        this.config = config;
    }

    public Result run(String flowName) {
        return run(flowName, Map.of(), null, false);
    }

    public Result run(String flowName, Map<String, String> env) {
        return run(flowName, env, null, false);
    }

    public Result run(String flowName, Map<String, String> env, Long timeoutOverrideMs, boolean reinstallDriver) {
        Path flowPath = Paths.get(config.mobile().flowsDir(), flowName + ".yaml").toAbsolutePath();
        long timeoutMs = timeoutOverrideMs != null ? timeoutOverrideMs : config.mobile().timeoutMs();
        String bin = config.mobile().maestroBin() != null ? config.mobile().maestroBin() : "maestro";

        List<String> cmd = new ArrayList<>();
        cmd.add(bin);
        if (config.mobile().deviceId() != null) {
            cmd.add("--device");
            cmd.add(config.mobile().deviceId());
        }
        cmd.add("test");
        cmd.add(flowPath.toString());
        if (reinstallDriver) {
            cmd.add("--reinstall-driver");
        }
        if (env != null) {
            env.forEach((k, v) -> {
                cmd.add("--env");
                cmd.add(k + "=" + v);
            });
        }

        boolean streamOutput = !"0".equals(System.getenv("DEBUG_MAESTRO")) && !"false".equals(System.getenv("DEBUG_MAESTRO"));

        try {
            Process proc = new ProcessBuilder(cmd).redirectErrorStream(true).start();
            StringBuilder output = new StringBuilder();
            Thread pump = pumpStream(proc, output, streamOutput);

            boolean exited = proc.waitFor(timeoutMs, TimeUnit.MILLISECONDS);
            if (!exited) {
                proc.destroy();
                if (!proc.waitFor(2, TimeUnit.SECONDS)) {
                    proc.destroyForcibly();
                }
                pump.join(1000);
                return new Result(false, output + "\nTimeout exceeded", -1);
            }
            pump.join(1000);
            int exitCode = proc.exitValue();
            boolean failed = exitCode != 0 || output.toString().contains("Flow Failed");
            return new Result(!failed, output.toString(), exitCode);
        } catch (IOException | InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Failed to run Maestro flow " + flowName, e);
        }
    }

    public Result runOrThrow(String flowName) {
        return runOrThrow(flowName, Map.of());
    }

    public Result runOrThrow(String flowName, Map<String, String> env) {
        Result result = run(flowName, env, null, false);
        if (!result.success()) {
            if (DRIVER_CRASH.matcher(result.output()).find()) {
                Result retry = run(flowName, env, null, true);
                if (retry.success()) {
                    return retry;
                }
                result = retry;
            }
            throw new RuntimeException("Maestro flow \"" + flowName + "\" failed:\n" + result.output());
        }
        return result;
    }

    public byte[] takeScreenshot(String name) {
        String androidHome = System.getenv("ANDROID_HOME");
        String adbBin = androidHome != null
                ? Paths.get(androidHome, "platform-tools", "adb").toString()
                : "adb";

        List<String> cmd = new ArrayList<>();
        cmd.add(adbBin);
        if (config.mobile().deviceId() != null) {
            cmd.add("-s");
            cmd.add(config.mobile().deviceId());
        }
        cmd.add("exec-out");
        cmd.add("screencap");
        cmd.add("-p");

        try {
            Process proc = new ProcessBuilder(cmd).start();
            byte[] png = proc.getInputStream().readAllBytes();
            proc.waitFor(30, TimeUnit.SECONDS);

            Path outDir = Paths.get("reports/screenshots");
            Files.createDirectories(outDir);
            Path outFile = outDir.resolve(name + ".png");
            Files.write(outFile, png);
            return png;
        } catch (IOException | InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Failed to capture mobile screenshot", e);
        }
    }

    private static Thread pumpStream(Process proc, StringBuilder sink, boolean streamToStderr) {
        Thread thread = new Thread(() -> {
            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(proc.getInputStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    sink.append(line).append('\n');
                    if (streamToStderr) {
                        System.err.println(line);
                    }
                }
            } catch (IOException ignored) {
                // process ended / stream closed
            }
        }, "maestro-stdout");
        thread.setDaemon(true);
        thread.start();
        return thread;
    }
}
